examples/src/rescue_raps/mod.rs

```diff
@@ -0,0 +1,167 @@
+        permuted_seeds.push(seeds[0]);
+        permuted_seeds.push(seeds[1]);
+
+        // compute the sequence of hashes using external implementation of Rescue hash
+        let now = Instant::now();
+        let result = compute_permuted_hash_chains(&seeds, &permuted_seeds);
+        debug!(
+            "Computed two permuted chains of {} Rescue hashes in {} ms",
+            chain_length,
+            now.elapsed().as_millis(),
+        );
+
+        RescueRapsExample {
+            options,
+            chain_length,
+            seeds,
+            permuted_seeds,
+            result,
+        }
+    }
+}
+
+// EXAMPLE IMPLEMENTATION
+// ================================================================================================
+
+impl Example for RescueRapsExample {
+    fn prove(&self) -> StarkProof {
+        // generate the execution trace
+        debug!(
+            "Generating proof for computing a chain of {} Rescue hashes\n\
+            ---------------------",
+            self.chain_length
+        );
+
+        // create a prover
+        let prover = RescueRapsProver::new(self.options.clone());
+
+        // generate the execution trace
+        let now = Instant::now();
+        let trace = prover.build_trace(&self.seeds, &self.permuted_seeds, self.result);
+        let trace_length = trace.length();
+        debug!(
+            "Generated execution trace of {} registers and 2^{} steps in {} ms",
+            trace.width(),
+            log2(trace_length),
+            now.elapsed().as_millis()
+        );
+
+        // generate the proof
+        prover.prove(trace).unwrap()
+    }
+
+    fn verify(&self, proof: StarkProof) -> Result<(), VerifierError> {
+        let pub_inputs = PublicInputs {
+            result: self.result,
+        };
+        winterfell::verify::<RescueRapsAir>(proof, pub_inputs)
+    }
+
+    fn verify_with_wrong_inputs(&self, proof: StarkProof) -> Result<(), VerifierError> {
+        let pub_inputs = PublicInputs {
+            result: [self.result[1], self.result[0]],
+        };
+        winterfell::verify::<RescueRapsAir>(proof, pub_inputs)
+    }
+}
+
+// HELPER FUNCTIONS
+// ================================================================================================
+
+fn absorb(state: &mut [BaseElement; STATE_WIDTH], values: &[BaseElement; 2]) {
+    state[0] += values[0];
+    state[1] += values[1];
+    for i in 0..NUM_HASH_ROUNDS {
+        rescue::apply_round(state, i);
+    }
+}
+
+fn compute_permuted_hash_chains(
+    seeds: &[[BaseElement; 2]],
+    permuted_seeds: &[[BaseElement; 2]],
+) -> [[BaseElement; STATE_WIDTH]; 2] {
+    let mut state = [BaseElement::ZERO; STATE_WIDTH];
+    let mut permuted_state = [BaseElement::ZERO; STATE_WIDTH];
+
+    // Start the hash chain
+    for (seed, permuted_seed) in seeds.iter().zip(permuted_seeds) {
+        absorb(&mut state, seed);
+        absorb(&mut permuted_state, permuted_seed);
+    }
+
+    [state, permuted_state]
```

Comment:
I'm wondering if we should return just the first two elements of each state here. For example, something like:

```Rust
[[state[0], state[1]], [permuted_state[0], permuted_state[1]]]
```

Not hugely important for this example, but in theory, having the entire hasher state be a public input would make the hash chain invertible.